### PR TITLE
Move root config fields to `[general]` section

### LINKED ISSUE #7906: Move `GENERAL` options into `[general]` section
The fact that this section should be at the top is confusing for some users, thus make a dedicated section for it and write an automatic migration rules to move things into it.

Certain fields, like `shell` could be moved into other sections, like `[terminal]`, however things like `live_config_reload` can not, and are enabled for the whole instance.

Maybe `[instance]` section could be used instead, but it feels not general enough.

### PR DESCRIPTION
**Pull request by chrisduerr**
Some users struggle with TOML, since root options must always be at the top of the file, since they're otherwise associated with the last table.

To avoid misunderstandings, all root-level fields have been removed. A new `general` section was added to allow housing configuration options that do not fit into any more specific groups.

Closes #7906.

### CODE REVIEW

File: alacritty/src/config/general.rs
Comment (chrisduerr) [inline alacritty/src/config/general.rs:26]: Switched this from `cfg(unix)` to `allow(unused)` because otherwise you'd get warnings on Windows with this in your config. We don't need to change the old one since that will warn for deprecation anyway. Realistically this should have never gotten added with a `cfg` attribute.

File: alacritty/src/cli.rs
Comment (kchibisov) [inline alacritty/src/cli.rs:94]: You can remove `{}` around it now.

File: alacritty/src/migrate.rs
Comment (kchibisov) [inline alacritty/src/migrate.rs:180]: Could we have a way to use `alacritty migrate` for `toml` to `toml` transition to reduce user amount of work? I think it should be relatively trivial and we already discussed something like that in the past?

Though, I won't block this PR on this small thing, just would be nice for the next release and other future releases.

File: alacritty/src/config/ui_config.rs
Comment (kchibisov) [inline alacritty/src/config/ui_config.rs:110]: Maybe `allow(unused)` here as well?

**Review by kchibisov — CHANGES_REQUESTED** [review]

File: alacritty/src/migrate.rs
Comment (chrisduerr) [reply alacritty/src/migrate.rs:180]: > Could we have a way to use alacritty migrate for toml to toml transition to reduce user amount of work? I think it should be relatively trivial and we already discussed something like that in the past?

I have considered doing this but decided against it because rewriting the toml config is going to mess up formatting, things like keybindings, nuke comments. I think to do that in a reasonable manner we'd have to preserve that stuff and I don't know how.

Do you think this is feasible?

File: alacritty/src/config/ui_config.rs
Comment (chrisduerr) [reply alacritty/src/config/ui_config.rs:110]: I intentionally didn't change it here because it's deprecated anyway. If you have problems with this warning it's just one more reason to migrate away.

File: alacritty/src/migrate.rs
Comment (kchibisov) [reply alacritty/src/migrate.rs:180]: I'm pretty sure you can do that with https://docs.rs/toml_edit/latest/toml_edit/ which we already use.

it explicitly states that everything is _preserved_ and just added and that's how `cargo add` works, which also adds things to toml, but preserves everything inside of it.

File: alacritty/src/config/ui_config.rs
Comment (kchibisov) [reply alacritty/src/config/ui_config.rs:110]: yeah, I don't mind much for deprecated that much.

File: alacritty/src/migrate.rs
Comment (kchibisov) [reply alacritty/src/migrate.rs:180]: Just thinking about it, your issue is the comment above the key we're about to move? Maybe https://docs.rs/toml_edit/latest/toml_edit/struct.Formatted.html kind of captures it?

Comment (chrisduerr) [reply alacritty/src/migrate.rs:180]: I'll investigate it. Might as well do so with this PR.

File: alacritty/src/migrate/yaml.rs
Comment (chrisduerr) [inline alacritty/src/migrate/yaml.rs:27]: In theory it would probably be possible to simplify this even more, by removing all the YAML import code and just moving the YAML file immediately and then performing a TOML migration, however that will likely change the way output is done and it makes dry-runs and removal of those files in error scenarios more difficult.

So considering that YAML is on its way out anyway and we'll remove the migration for it in the future, I don't think it's worth it to risk breaking anything here.

**Comment by chrisduerr** [comment]
@kchibisov Migration seems to work great, I'd just rebase it as two PRs and it should be good to go I think.

File: alacritty/src/migrate/mod.rs
Comment (kchibisov) [inline alacritty/src/migrate/mod.rs:83]: I think we should indicate error here, so it's less confusing.

File: alacritty/src/migrate/mod.rs
Comment (kchibisov) [inline alacritty/src/migrate/mod.rs:71]: I'd probably log the `path` here along the error, but it doesn't really matter.

**Review by kchibisov — CHANGES_REQUESTED** [review]
I think we should suggest for deprecated fields to run `alacritty migrate` to automate the process.

Can do something like that to move original users' comments.

```diff
diff --git a/alacritty/src/migrate/mod.rs b/alacritty/src/migrate/mod.rs
index ccf4cedf..10d8ce23 100644
--- a/alacritty/src/migrate/mod.rs
+++ b/alacritty/src/migrate/mod.rs
@@ -156,27 +156,49 @@ fn migrate_imports(
 /// Move a TOML value from one map to another.
 fn move_value(document: &mut DocumentMut, origin: &[&str], target: &[&str]) -> Result<(), String> {
     // Find and remove the original item.
-    let mut origin_item = document.as_item_mut();
+    let (mut origin_key, mut origin_item) = (None, document.as_item_mut());
     for element in origin {
-        origin_item = &mut origin_item[element];
+        let table = match origin_item.as_table_like_mut() {
+            Some(table) => table,
+            None => panic!("Moving from unsupported TOML structure"),
+        };

-        // Abort early if item doesn't exist.
-        if origin_item.is_none() {
-            return Ok(());
-        }
+        let (key, item) = match table.get_key_value_mut(element) {
+            Some((key, item)) => (key, item),
+            None => return Ok(()),
+        };
+
+        dbg!(&key);
+        origin_key = Some(key);
+        origin_item = item;

         // Ensure no empty tables are left behind.
         if let Some(table) = origin_item.as_table_mut() {
             table.set_implicit(true)
         }
     }
+
+    let origin_key_decor =
+        origin_key.map(|key| (key.leaf_decor().clone(), key.dotted_decor().clone()));
     let origin_item = mem::replace(origin_item, Item::None);

     // Create all dependencies for the new location.
     let mut target_item = document.as_item_mut();
     for (i, element) in target.iter().enumerate() {
+        let table = match target_item.as_table_like_mut() {
+            Some(table) => table,
+            None => panic!("Moving into unsupported TOML structure"),
+        };
+
         if i + 1 == target.len() {
-            target_item[element] = origin_item;
+            table.insert(element, origin_item);
+            // Move original key decorations.
+            if let Some((leaf, dotted)) = origin_key_decor {
+                let mut key = table.key_mut(element).unwrap();
+                *key.leaf_decor_mut() = leaf;
+                *key.dotted_decor_mut() = dotted;
+            }
+
             break;
         } else {
             // Create missing parent tables.
@@ -249,6 +271,9 @@ mod tests {
     #[test]
     fn move_values() {
         let input = r#"
+# This is a root_value.
+#
+# Use it with care.
 root_value = 3

 [table]
@@ -278,6 +303,10 @@ not_moved = 9
 new_name = 5

 [new_table]
+
+# This is a root_value.
+#
+# Use it with care.
 root_value = 3
         "#;

```

File: alacritty/src/migrate/mod.rs
Comment (chrisduerr) [reply alacritty/src/migrate/mod.rs:83]: It's possible to have migrations that have both yaml and toml files though. So it's not technically an error and there might be other things done by migrate in the future which aren't renames.

That said, I doubt this option was ever used outside of testing. So I'll settle for a warning for now and we can probably remove it together with the yaml stuff.

File: alacritty/src/migrate/mod.rs
Comment (chrisduerr) [reply alacritty/src/migrate/mod.rs:71]: We already do in our error handling:

```
Configuration file migration failed:
    "/tmp/test": missing file extension
```

**Comment by chrisduerr** [comment]
> I think we should suggest for deprecated fields to run alacritty migrate to automate the process.

I've considered this but decided against it since that would bloat our warnings. Do you have a specific suggestion that would keep our warnings short while still providing helpful instructions?

**Comment by kchibisov** [comment]
The `deprecated` ones are not that big though, so 3-4 more words won't do much. Or do you mean that you don't want manually to specify all of that? Could be auto-added by config crate for all deprecated fields.

**Comment by chrisduerr** [comment]
Our existing messages are already quite long:

```
Config warning: ipc_socket has been deprecated; use general.ipc_socket instead
```

To include `alacritty migrate` would make them significantly longer:

```
Config warning: ipc_socket has been deprecated; run `alacritty migrate` or use general.ipc_socket instead
```

The current message with this pretty short config option is already 78 columns wide. Generally I would like to stay at or below 80 to ensure we're not broken across multiple lines on most setups. With explicitly pointing out `alacritty migrate` in every warning, I don't see how we'd go below 100 columns, which is pretty wide.

**Comment by kchibisov** [comment]
I was thinking more like

```
Config warning: ipc_socket has been deprecated; use general.ipc_socket instead
Run alacritty migrate to automatically resolve it
```

So on a separate line.

**Comment by chrisduerr** [comment]
I've added it on a separate line. I'm not sure I love it but it might help some people?

**Review by kchibisov — APPROVED** [review]

**Comment by dann-merlin** [comment]
@chrisduerr This was great for me! Thanks a lot. Little rant, but I'm using hyprland and whenever I update that, there's a bunch of config options that have been renamed/moved/removed and I need to go hunting in the commits or something.
Recently I found a comment on github from the maintainer "Oh, I announced that on my Discord"??? Like I want to be a social media follower or something...
I just want to use the software and not worry about something breaking because I did an update.

So when I saw the message
```
Config warning: import has been deprecated; use general.import instead
Use 'alacritty migrate' to automatically resolve it
```
and actually ran `alacritty migrate` and everything worked, that was really refreshing.
I felt like I really need to give some props to you. Nicely done and thanks!

**Comment by chrisduerr** [comment]
I think @kchibisov deserves credit for putting that message in the message bar.

**Comment by dann-merlin** [comment]
Fair! Thanks to both of you! @kchibisov @chrisduerr

**Pull request merged**
